pkg/reconciler/tenancy/initialization/apibinder/apiset_controller.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibinder
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	apisinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	initializationv1alpha1 "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/apis/initialization/v1alpha1"
+	initializationclient "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/clientset/versioned"
+	initializationinformer "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/informers/externalversions/initialization/v1alpha1"
+	initializationlister "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/listers/initialization/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefinition"
+)
+
+const (
+	apiSetControllerName = "kcp-apiset-controller"
+	byExport             = apiSetControllerName + "-byExport"
+)
+
+type CreateAPIDefinitionFunc func(apiResourceSchema *apisv1alpha1.APIResourceSchema, version string, identityHash string) (apidefinition.APIDefinition, error)
+
+// NewAPISetValidator returns a new controller which watches APISets ensures they are valid.
+func NewAPISetValidator(
```

Comment:
(no action required) mismatch between file name and struct - apiset_controller.go vs APISetValidator